lldb/source/Target/SwiftLanguageRuntimeNames.cpp

```diff
@@ -263,69 +291,86 @@ class ThreadPlanStepInAsync : public ThreadPlan {
-    auto reg_ctx = frame_sp->GetRegisterContext();
-
-    // swift_task_switch(AsyncTask *, ExecutorRef, ExecutorRef)
-    constexpr auto task_regnum = LLDB_REGNUM_GENERIC_ARG1;
-    auto task_reg = reg_ctx->ConvertRegisterKindToRegisterNumber(
-        RegisterKind::eRegisterKindGeneric, task_regnum);
-    auto task_ptr = reg_ctx->ReadRegisterAsUnsigned(task_reg, 0);
-    if (!task_ptr) {
+  BreakpointSP CreateAsyncBreakpoint(Thread &thread) {
+    // The signature for `swift_task_switch` is as follows:
+    //   SWIFT_CC(swiftasync)
+    //   void swift_task_switch(
+    //     SWIFT_ASYNC_CONTEXT AsyncContext *resumeContext,
+    //     TaskContinuationFunction *resumeFunction,
+    //     ExecutorRef newExecutor);
+    //
+    // The async context given as the first argument is not passed using the
+    // calling convention's first register, it's passed in the platform's async
+    // context register. This means the `resumeFunction` parameter uses the
+    // first ABI register (ex: x86-64: rdi, arm64: x0).
+    auto reg_ctx = thread.GetStackFrameAtIndex(0)->GetRegisterContext();
+    constexpr auto resume_fn_regnum = LLDB_REGNUM_GENERIC_ARG1;
+    auto resume_fn_reg = reg_ctx->ConvertRegisterKindToRegisterNumber(
+        RegisterKind::eRegisterKindGeneric, resume_fn_regnum);
+    auto resume_fn_ptr = reg_ctx->ReadRegisterAsUnsigned(resume_fn_reg, 0);
+    if (!resume_fn_ptr) {
       return {};
     }
 
-    auto process_sp = thread.GetProcess();
-    const auto sizeof_pointer = process_sp->GetAddressByteSize();
-    auto run_job_ptr = task_ptr;
-    if (sizeof_pointer == 8) {
-      run_job_ptr += taskOffsetOfRunJob64;
-    } else if (sizeof_pointer == 4) {
-      run_job_ptr += taskOffsetOfRunJob32;
-    }
+    auto &target = thread.GetProcess()->GetTarget();
+    auto breakpoint_sp = target.CreateBreakpoint(resume_fn_ptr, true, false);
+    breakpoint_sp->SetBreakpointKind("async-step");
+    breakpoint_sp->SetCallback(AsyncBreakpointHit, this, true);
+    return breakpoint_sp;
+  }
 
-    Status status;
-    auto fn_ptr = process_sp->ReadPointerFromMemory(run_job_ptr, status);
-    if (status.Fail()) {
-      return {};
+  static bool AsyncBreakpointHit(void *baton, StoppointCallbackContext *context,
+                                 lldb::user_id_t break_id,
+                                 lldb::user_id_t break_loc_id) {
+    auto *plan = reinterpret_cast<ThreadPlanStepInAsync *>(baton);
+    plan->SetTID(context->exe_ctx_ref.GetThreadSP()->GetID());
+    plan->m_breakpoint_async_context = GetAsyncContext(context->exe_ctx_ref.GetFrameSP());
+    return true;
+  }
+
+  static uint64_t GetAsyncContext(lldb::StackFrameSP frame_sp) {
+    auto reg_ctx_sp = frame_sp->GetRegisterContext();
+    auto arch = reg_ctx_sp->CalculateTarget()->GetArchitecture();
+    int async_context_regnum = 0;
+    if (arch.GetMachine() == llvm::Triple::x86_64) {
+      async_context_regnum = dwarf_r14_x86_64;
```

Comment:
It would be nice to consolidate the knowledge about r14/r22 into a single place, too.